### PR TITLE
Use source for errors instead of path

### PR DESCRIPTION
**Pull request by pmukerji**
The JSON API format specifies using a `source` key to reference the source
of the error in Error Objects (http://jsonapi.org/format/#error-objects).

Currently the key `path` is being used.

At the moment we are also getting a `path` that looks like: e.g. `/email` for an attribute `email`

The JSON API format specifies:

> `pointer`: a JSON Pointer [RFC6901] to the associated entity in the request document [e.g. "/data" for a primary data object, or "/data/attributes/title" for a specific attribute].

Should we update to specify that it is an attribute like: `/data/attributes/email`?

also I wasn't clear if we should also have `pointer` specified, to make it clear it is a JSON Pointer instead of a parameter as a string.

### CODE REVIEW

**Comment by lgebhardt** [comment]
Good catch. I didn't notice those changes to spec. As I read it `source` could either contain the source of the error as a string or a sub key of either `pointer` or `parameter` which has the details. I'd say we should use these optional keys as they remove some ambiguity.

**Comment by pmukerji** [comment]
Cool I've added the `pointer` sub key, but still need to check how best to add the path to a specific attribute:

`/data/attributes/email` instead of `/email`. I'm wondering what the pointer should be for an invalid relationship? 

e.g. for creating a `BookResource` and specifying a (`has_one`) `LibraryResource` that doesn't exist

``` json
{
  "data": {
    "type": "books",
    "relationships": {
      "library": {
        "data": { "type": "libraries", "id": "4" }
      }
    }
}
```

``` json
{
  "errors": [
    {
      "title": "library - can't be blank",
      "detail": "can't be blank",
      "id": null,
      "href": null,
      "code": 100,
      "source": "/library",
      "links": null,
      "status": "unprocessable_entity"
    }
  ]
}
```

Here I guess `source` should be something like: `/data/relationships/library` or `/data/relationships/library/data/id` ?

**Comment by pmukerji** [comment]
@lgebhardt - Took a shot at implementing the more specific error pointer

**Comment by lgebhardt** [comment]
Thanks. Other than my one comment above I think this is ready to merge.

**Comment by pmukerji** [comment]
Great, I initialized with just the resource now, and did some minor refactoring

File: lib/jsonapi/exceptions.rb [hunk 1]
```diff
@@ -305,16 +307,26 @@ def format_key(key)
...
-              JSONAPI::Error.new(code: JSONAPI::VALIDATION_ERROR,
-                                 status: :unprocessable_entity,
-                                 title: "#{format_key(element[0])} - #{message}",
-                                 detail: message,
-                                 path: "/#{element[0]}")
-            end
-          )
+        error_messages.map do |attr_key, messages|
+          messages.map { |message| json_api_error(attr_key, message) }
+        end.flatten
```
Comment (fphilipe) [inline lib/jsonapi/exceptions.rb]: Could replace the first `map` with `flat_map` and get rid of the `flatten`.

Comment (pmukerji) [inline lib/jsonapi/exceptions.rb]: :+1: thanks! Updated.

**Comment by lgebhardt** [comment]
Thanks again!

**Comment by pmukerji** [comment]
No problem! Thanks for the quick merge and release

Just a small thing from the release notes:

```
Errors now use path instead of source parameter, as per the JSON API spec.
```

I guess path and source are reversed here. Not sure if you can edit the release notes, but it could read:

```
Errors now use source instead of path parameter, as per the JSON API spec.
```

**Comment by lgebhardt** [comment]
@pmukerji Thanks. Updated.

**Pull request merged**